### ISSUE TITLE
endRecording().toImage may provide empty data on web

### ISSUE THREAD

**Issue by DavidKuennen** [flutter/flutter#40960]
It seems to me, that `canvas.drawImage` and `canvas.drawImageRect` isn't recorded properly in some cases. 

You can see it in the demo repo if you click the button very fast.

DEMO REPO: https://github.com/DavidKuennen/flutter-to-image-error-demo

How to reproduce in Demo-Repo:
![Documentation Image](https://storage.googleapis.com/devsnap-static-assets/ezgif.com-video-to-gif.gif)

```
flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.9.1+hotfix.2, on Microsoft Windows [Version
    10.0.18362.356], locale en-US)
 
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.2)
[√] VS Code, 64-bit edition (version 1.38.0)
[√] Connected device (1 available)

• No issues found!
```

Related to: https://github.com/flutter/flutter/issues/40244

**Comment by escamoteur** [comment]
Hi have you tried if this is also the case if you use `master channel`?

**Comment by DavidKuennen** [comment]
It also happens on `master channel`. I'm unable to test it on iOS though.

**Comment by DavidKuennen** [comment]
My current workaround (I know, it's hacky as hell, but I've got no other choice): 

```dart
Future<ByteData> get rendered async {
  for(var i = 0; i < 10; i++) {
    var byteData = await getByteData();
    if(byteData == null) {
      return null;
    } else if(byteData.lengthInBytes > originalSize / 4) {
      return byteData;
    }
    await Future.delayed(const Duration(milliseconds: 10));
  }
  return null;
}
```

**Comment by a27919478** [comment]
I also encountered this problem.

**Comment by a27919478** [comment]
I also encountered this problem.

**Comment by flbaue** [comment]
I have this issue as well. But so far only on actuall iPhone devices and not the simulator.

**Comment by Forsaknd** [comment]
Same here, happens rarely but it does occur on both Android and iOS (version does not seem to matter tried most major Android versions after lolipop). In my app users upload images and the "image page" closes after successful upload so sometimes a blank image is uploaded without the user noticing. Worked around it by users having to leave the page manually, which is less convenient, so would very much like this to get looked into. Altough i bet its hard to debug as it happens so rarely.

**Comment by csells** [comment]
@DavidKuennen is this still happening against Flutter 1.12?

**Comment by iapicca** [comment]
@csells 
Just made a copy-paste-able [sample](https://dartpad.dev/9b2a8c4dc00a88f7f2f1c3eb9d0815d5)
on web doesn't work at all
while on iOS and especially on android behave weirdly.
I hope it can help

**Comment by TahaTesser** [comment]
Hi @DavidKuennen
Using repo, It consistently provides data when I tap several times
https://github.com/DavidKuennen/flutter-to-image-error-demo

![ezgif com-resize (4)](https://user-images.githubusercontent.com/48603081/94037876-6f9a2d00-fde3-11ea-879e-c97ca49ac87a.gif)



Are you still experiencing this issue? 

If yes, can you please provide your `flutter doctor -v` with the latest `stable` and `flutter run --verbose` logs, and a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)
Thank you

**Comment by TahaTesser** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

**Comment by iapicca** [comment]
@TahaTesser it still doesn't work on web [sample](https://github.com/flutter/flutter/issues/40960#issuecomment-591909798) ([source](https://github.com/flutter/flutter/issues/40960#issuecomment-591909798))

**Comment by TahaTesser** [comment]
<details> 
<summary>complete code sample</summary> 

```dart
import 'dart:async';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(
        home: MyHomePage(),
      );
}

class MyHomePage extends StatelessWidget {
  static const String _url =
      'https://flutter.dev/assets/homepage/carousel/slide_2-bg-ec22a39fb182cb26ec8fb88b30064bc08707d4e87b0e7415c2f077c20d3e6102.jpg';
  static NetworkImage get _networkImage => NetworkImage(_url);
  final NetworkImageDecoder _decoder =
      NetworkImageDecoder(image: _networkImage);

  @override
  Widget build(BuildContext context) => FutureBuilder<ui.Image>(
        future: _decoder.uiImage,
        builder: (BuildContext context, AsyncSnapshot snapshot) =>
            !snapshot.hasData
                ? const Center(
                    child: const CircularProgressIndicator(),
                  )
                : PainterWidget(
                    image: snapshot.data,
                  ),
      );
}

class PainterWidget extends StatefulWidget {
  final ui.Image image;
  const PainterWidget({
    this.image,
  });

  @override
  _PainterWidget createState() => _PainterWidget();
}

class _PainterWidget extends State<PainterWidget> {
  CustomPainter _painter;
  PainterDecoder _decoder;
  StreamController<Uint8List> _controller;

  @override
  void initState() {
    super.initState();
    _controller = StreamController<Uint8List>();
    _painter = MyCustomPainter(image: widget.image);
    _decoder = PainterDecoder(
      painter: _painter,
      size: Size(
        widget.image.width.toDouble(),
        widget.image.height.toDouble(),
      ),
    );
  }

  @override
  void dispose() {
    _controller.close();
    super.dispose();
  }

  void _pump() async =>
      _decoder.uint8list.then((value) => _controller.sink.add(value));

  @override
  Widget build(BuildContext context) => SafeArea(
        child: Scaffold(
          body: Stack(
            children: <Widget>[
              Align(
                alignment: Alignment.topCenter,
                child: CustomPaint(
                  painter: _painter,
                ),
              ),
              Align(
                alignment: Alignment.bottomCenter,
                child: StreamBuilder<Uint8List>(
                    stream: _controller.stream,
                    builder: (BuildContext context, AsyncSnapshot snapshot) =>
                        !snapshot.hasData
                            ? const Center(
                                child: const CircularProgressIndicator(),
                              )
                            : Image.memory(snapshot.data)),
              ),
            ],
          ),
          floatingActionButton: FloatingActionButton(
            child: Icon(Icons.star),
            onPressed: _pump,
          ),
        ),
      );
}

class NetworkImageDecoder {
  final NetworkImage image;
  const NetworkImageDecoder({this.image});

  Future<ImageInfo> get imageInfo async {
    final Completer<ImageInfo> completer = Completer();
    image.resolve(ImageConfiguration()).addListener(
          ImageStreamListener(
            (ImageInfo info, bool _) => completer.complete(info),
          ),
        );
    return await completer.future;
  }

  Future<ui.Image> get uiImage async {
    final ImageInfo _info = await imageInfo;
    return _info.image;
  }
}

class MyCustomPainter extends CustomPainter {
  final ui.Image image;
  const MyCustomPainter({
    this.image,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final Paint _paint = Paint()..isAntiAlias = true;
    canvas
      ..drawImageRect(image, Rect.fromLTRB(0, 0, 300, 300),
          Rect.fromLTRB(0, 0, 300, 300), _paint)
      ..drawLine(Offset(0, 0), Offset(300, 300), _paint);
  }

  @override
  bool shouldRepaint(MyCustomPainter oldDelegate) => true;
}

class PainterDecoder {
  final CustomPainter painter;
  final Size size;
  const PainterDecoder({
    this.painter,
    this.size,
  });

  ui.Picture get picture {
    final ui.PictureRecorder recorder = ui.PictureRecorder();
    final Canvas canvas = Canvas(recorder);
    painter.paint(canvas, size);
    return recorder.endRecording();
  }

  Future<ui.Image> get image =>
      picture.toImage(size.width.toInt(), size.height.toInt());

  Future<ByteData> get imageData async {
    final ui.Image _image = await image;
    return _image.toByteData(format: ui.ImageByteFormat.png);
  }

  Future<Uint8List> get uint8list async {
    final ByteData _data = await imageData;
    return _data.buffer.asUint8List();
  }
}

```

</details>

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.24.0-4.0.pre.57, on Mac OS X 10.15.7 19H2
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-4.0.pre.57 at /Users/tahatesser/Code/flutter_master
    • Framework revision fdac8a9d1f (14 hours ago), 2020-10-25 19:47:02 -0400
    • Engine revision d4a3c9c471
    • Dart version 2.11.0 (build 2.11.0-256.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0.rc.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (5 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E            • ios
      • iOS 14.1
    • iPhone 12 (mobile)     • 318957C3-E4BB-4B35-A3C7-14F515B0E665 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-1 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64
      • Mac OS X 10.15.7 19H2 darwin-x64
    • Web Server (web)       • web-server                           •
      web-javascript • Flutter Tools
    • Chrome (web)           • chrome                               •
      web-javascript • Google Chrome 86.0.4240.111

• No issues found!
```

</details>

**Comment by huycozy** [comment]
Hi @DavidKuennen, this issue seems to have been fixed on the latest master channel `3.6.0` as I cannot reproduce it there. See the demo below.

<details>
<summary>Demo</summary>

| stable 3.3.9 | master 3.6.0 |
| --------------- | --------------- |
<video src="https://user-images.githubusercontent.com/104349824/203774986-7127a2d2-2ed2-4b59-bf62-c50134af622e.mov"/> | <video src="https://user-images.githubusercontent.com/104349824/203775171-b57b139b-67c0-4906-ab68-cb4eebf8b76d.mov"/>

</details>

<details>
<summary>Sample code (null safety supported)</summary>

```dart
import 'dart:async';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(
    home: MyHomePage(),
  );
}

class MyHomePage extends StatelessWidget {
  static const String _url =
      'https://cdn.dribbble.com/users/1622791/screenshots/11174104/flutter_intro.png';
  static NetworkImage get _networkImage => NetworkImage(_url);
  final NetworkImageDecoder _decoder =
  NetworkImageDecoder(image: _networkImage);

  @override
  Widget build(BuildContext context) => FutureBuilder<ui.Image>(
    future: _decoder.uiImage,
    builder: (BuildContext context, AsyncSnapshot snapshot) =>
    !snapshot.hasData
        ? const Center(
      child: const CircularProgressIndicator(),
    )
        : PainterWidget(
      image: snapshot.data,
    ),
  );
}

class PainterWidget extends StatefulWidget {
  final ui.Image image;
  const PainterWidget({
    required this.image,
  });

  @override
  _PainterWidget createState() => _PainterWidget();
}

class _PainterWidget extends State<PainterWidget> {
  late CustomPainter _painter;
  late PainterDecoder _decoder;
  late StreamController<Uint8List> _controller;

  @override
  void initState() {
    super.initState();
    _controller = StreamController<Uint8List>();
    _painter = MyCustomPainter(image: widget.image);
    _decoder = PainterDecoder(
      painter: _painter,
      size: Size(
        widget.image.width.toDouble(),
        widget.image.height.toDouble(),
      ),
    );
  }

  @override
  void dispose() {
    _controller.close();
    super.dispose();
  }

  void _pump() async =>
      _decoder.uint8list.then((value) {
        if(value!=null) _controller.sink.add(value);
      });

  @override
  Widget build(BuildContext context) => SafeArea(
    child: Scaffold(
      body: Stack(
        children: <Widget>[
          Align(
            alignment: Alignment.topCenter,
            child: CustomPaint(
              painter: _painter,
            ),
          ),
          Align(
            alignment: Alignment.bottomCenter,
            child: StreamBuilder<Uint8List>(
                stream: _controller.stream,
                builder: (BuildContext context, AsyncSnapshot snapshot) =>
                !snapshot.hasData
                    ? const Center(
                  child: const CircularProgressIndicator(),
                )
                    : Image.memory(snapshot.data)),
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.star),
        onPressed: _pump,
      ),
    ),
  );
}

class NetworkImageDecoder {
  final NetworkImage image;
  const NetworkImageDecoder({required this.image});

  Future<ImageInfo> get imageInfo async {
    final Completer<ImageInfo> completer = Completer();
    image.resolve(ImageConfiguration()).addListener(
      ImageStreamListener(
            (ImageInfo info, bool _) => completer.complete(info),
      ),
    );
    return await completer.future;
  }

  Future<ui.Image> get uiImage async {
    final ImageInfo _info = await imageInfo;
    return _info.image;
  }
}

class MyCustomPainter extends CustomPainter {
  final ui.Image image;
  const MyCustomPainter({
    required this.image,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final Paint _paint = Paint()..isAntiAlias = true;
    canvas
      ..drawImageRect(image, Rect.fromLTRB(0, 0, 300, 300),
          Rect.fromLTRB(0, 0, 300, 300), _paint)
      ..drawLine(Offset(0, 0), Offset(300, 300), _paint);
  }

  @override
  bool shouldRepaint(MyCustomPainter oldDelegate) => true;
}

class PainterDecoder {
  final CustomPainter painter;
  final Size size;
  const PainterDecoder({
    required this.painter,
    required this.size,
  });

  ui.Picture get picture {
    final ui.PictureRecorder recorder = ui.PictureRecorder();
    final Canvas canvas = Canvas(recorder);
    painter.paint(canvas, size);
    return recorder.endRecording();
  }

  Future<ui.Image> get image =>
      picture.toImage(size.width.toInt(), size.height.toInt());

  Future<ByteData?> get imageData async {
    final ui.Image _image = await image;
    return _image.toByteData(format: ui.ImageByteFormat.png);
  }

  Future<Uint8List?> get uint8list async {
    final ByteData? _data = await imageData;
    return _data?.buffer.asUint8List();
  }
}

```
</details>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.3.9, on macOS 13.0 22A380 darwin-x64, locale en-VN)
    • Flutter version 3.3.9 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b8f7f1f986 (31 hours ago), 2022-11-23 06:43:51 +0900
    • Engine revision 8f2221fbef
    • Dart version 2.18.5
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (3 available)
    • SM T225 (mobile) • R9JT3004VRJ • android-arm64  • Android 12 (API 31)
    • macOS (desktop)  • macos       • darwin-x64     • macOS 13.0 22A380 darwin-x64
    • Chrome (web)     • chrome      • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

```bash
[!] Flutter (Channel master, 3.6.0-7.0.pre.45, on macOS 13.0 22A380 darwin-x64, locale en-VN)
    • Flutter version 3.6.0-7.0.pre.45 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d0352f55db (3 hours ago), 2022-11-23 20:01:26 -0500
    • Engine revision 9b2895724c
    • Dart version 2.19.0 (build 2.19.0-424.0.dev)
    • DevTools version 2.19.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (3 available)
    • SM T225 (mobile) • R9JT3004VRJ • android-arm64  • Android 12 (API 31)
    • macOS (desktop)  • macos       • darwin-x64     • macOS 13.0 22A380 darwin-x64
    • Chrome (web)     • chrome      • web-javascript • Google Chrome 107.0.5304.110

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```
</details>

I'll be closing this issue since it has already been fixed on the master channel.
If you still experience this issue on the master channel, kindly leave a comment below and I will reopen this.
Thank you!
